Oracle.md

```diff
@@ -0,0 +1,198 @@
+The minimum and maximum values that can be attested to by an oracle should thus be interpreted as `max_value or more` and `min_value or less`.
+This enables contracting party to specify payouts for the overflow and underflow cases, and avoid having to use the refund path which in most cases would be unfair.
+More complex constructions where considered to handle these, but the simplicity of the currently specified approach is the reason that it was chosen.
+
+The oracle must separately provide an array of R values, one for each of the digit that will be signed, with the first value of the array being used for signing the leftmost digit. If the is-signed value is true, an additional R-value must be provided as the first element in this array, which will be used to sign the string "+" in case of an outcome with a positive value or zero value and the string "-" in case of an outcome with a negative one.
+In practice this array is provided as part of the [Oracle events](#Oracle-events) structure.
+
+When the outcome is not the same order of magnitude as the maximum possible outcome, leading zeros must be signed.
+
+#### Example: BTC/USD rate
+
+The following example illustrates how numerical decomposition can help reduce the number of required CETs while covering a large range of possible outcomes.
+
+Alice and Bob wish to enter into a DLC at time `t` when BTC/USD ~= $10000 with a maturity time of `t+1`.
+The oracle provides 6 R-values `R0-5`, meaning that it will sign six digits `D0-5`, enabling him to attest to outcomes in the range `[000000,999999]`:
+
+```
+base: 10
+isSigned: false
+RValues: [R0, R1, R2, R3, R4, R5]
+unit: dollars
+precision: 0
+```
+
+Alice and Bob create a DLC with "relevant" outcomes between $10000 and $10999.
+Below $10000 Alice gets everything, and above $10999 Bob gets everything.
+They use the `R0` and `R1` values to cover the cases where BTC/USD < $10000 using `D0=0` and `D1=0`.
+They use `R0`, `R1` and `R2` to cover the cases where BTC/USD > $10999 and <= $19999, using `D1=0`, `D1=1` and `D2 in [0-9]`.
+They use `R0` and `R1` to cover the cases where BTC/USD >= $20000 and < $99999, using `D0=0` and `D1 in [2,9]`.
+Finally, they use only `D0` to cover the cases where BTC/USD >= $100000 and <= $999999, using `D0 in [1-9]`.
+
+At maturity time, if the BTC/USD rate is less than $999999, the oracle signs each of the digits representing the outcome value using the digits' corresponding R value.
+For example, if the rate at maturity time is $20000, the oracle will sign the digits `0`, `2`, `0`, `0`, `0`, `0`.
+
+If the rate is greater than $999999, the oracle signs the digits `9`, `9`, `9`, `9`, `9`, `9`.
+
+### Serialization and signing of outcome values
+
+Every outcome value should be encoded as [a UTF-8 string with NFC normalization](./Messaging.md#Fundamental-types), hashed using sha256 before finally being signed by an oracle.
```

Comment:
I consider the hashing to be part of the signing algorithm. I think we just need to describe in another document a `sign(keypair, tag, message)` function which tags the message (e.g. tagged hash) and uses BIP340 `sign` internally. This document can then just say what the message and tag should be without mentioning tagged at all. Currently it's missing a tag e.g. `oracle/outcome`. Of course this doesn't exist yet so consider it just a comment.

I also think the message should not just be the raw outcome string (as discussed before) but should include the event id somehow. The format I am using is `{event_id}.{i}={outcome_str}` where `i` is the index (i.e. digit) of the signature. I don't really mind if we specify this later as it's an easy thing to change.

Reply:
I updated too mention that the encoded value is passed to "a signing algorithm". I haven't understood if you want the message format (`{event_id}.{i}={outcome_str}`) to be mentioned here or somewhere else?

Reply:
It can probably be here, just in a future revision.

Reply:
I opened an [issue](https://github.com/discreetlogcontracts/dlcspecs/issues/111) so that we can take care of this in a future PR